CRM/Core/DAO.php

```diff
@@ -1644,6 +1644,8 @@ public static function executeUnbufferedQuery(
    *   object that holds the results of the query
    *   NB - if this is defined as just returning a DAO phpstorm keeps pointing
    *   out all the properties that are not part of the DAO
+   *
+   * @throws \Civi\Core\Exception\DBQueryException
    */
   public static function &executeQuery(
```

Comment:
Just trying to think this through...  I do like having that clearly documented for someone who's reading the docblock.

But... `DBQueryException` is a "checked" exception (*insofar as it's not a `\RuntimeException`*), and PHPStorm will nag you to catch it. The thing about that... when searching `universe`, [there were only 2-3 small places](https://github.com/civicrm/civicrm-core/pull/25600#issuecomment-1437664866) which actually tried to catch an exception like this. But there are.. about 5,000 places in `universe/ext/` that call `executeQuery()`. So won't the tooling expect you to put `catch()` or `@throws` on all of ~5,000 of them (plus, for `@throws`, the same on every  downstream call)?

Reply:
The function actually throws 2 different types of exceptions - it also throws an exception if the validation is invalid.

We kind of have the problem in practice anyway - as all the places that call the api are required to catch `CRM_Core_Exception`.

I took a look & drupal DOES declare an exception on it's `prepare` so I guess drupal code has to catch those things.


https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Entity%21Query%21Sql%21Query.php/10

Reply:
> so I guess drupal code has to catch those things...

I started looking at their arrangement, and... it's a bit of a rabbit-hole. For example, here's a basic query:

```php
\Drupal::database()->query("SELECT id, example FROM {mytable}")->fetchAll();
```

The type-hints have a lot of abstraction there. Both in the immediate call chain (e.g. `abstract class Connection`, `interface StatementInterface`) and in the various `@throws`. On `Connection::query()`, it `@throws` one of these: 

```php
class DatabaseExceptionWrapper extends \RuntimeException implements DatabaseException {
```

So from the callers' POV, that's an unchecked exception. But if you want to catch it, then you should probably go for `DatabaseException` instead of `DatabaseExceptionWrapper`.

But there's more! In the class you linked, the `protected prepare()` declares a `@throws QueryException` but the `public execute()` doesn't declare any `@throws` at all. So they just hide that one...

I dunno, after looking at Drupal's example, I feel more confused. :roll_eyes: 

> .all the places that call the api or singleValueQuery are required to catch CRM_Core_Exception or document it (and if you catch or document CRM_Core_Exception that covers this.

Very good point!